### DISCUSSION TITLE
How to install ComfyUI Manager properly?

### DISCUSSION THREAD

**Discussion by gl-prout** [AbdBarho/stable-diffusion-webui-docker#686] [Q&A]
I have successfully installed ComfyUI and have cloned ComfyUI Manager's repo in `stable-diffusion-webui-docker/data/config/comfy/custom_nodes`

But when I restart ComfyUI, ComfyUI Manager responds with the module `git` missing, which makes sense, because ComfyUI Manager's python requirements are not installed within the image

Those are the said requirements as listed on their repo

```
GitPython
PyGithub
matrix-client==0.4.0
transformers
huggingface-hub>0.20
```

I do believe that I would need to modify either the `entrypoint.sh` or `Dockerfile` in `stable-diffusion-webui-docker/services/comfy` to add instructions to pip install those requirements, but I'm not sure where it would be the most appropriate, and the condition to see if ComfyUI Manager is there or not

**Comment by gl-prout** [comment]
I figured it out

I had to take a good look at `entrypoint.sh` to figure out that I had to create a `startup.sh` script in `stable-diffusion-webui-docker/data/config/comfy` and verify if the `requirements.txt` is present at said location or not

So, when I restart the container, it executes the `startup.sh` script

```bash
#!/bin/bash

if [ -f "/data/config/comfy/custom_nodes/ComfyUI-Manager/requirements.txt" ]; then
  pip install -r /data/config/comfy/custom_nodes/ComfyUI-Manager/requirements.txt
fi
```

However, I met another error message at container restart:

```
Traceback (most recent call last):
  File "/stable-diffusion/nodes.py", line 1810, in load_custom_node
    module_spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/data/config/comfy/custom_nodes/ComfyUI-Manager/__init__.py", line 6, in <module>
    from .glob import manager_server
  File "/data/config/comfy/custom_nodes/ComfyUI-Manager/glob/manager_server.py", line 18, in <module>
    import manager_core as core
ModuleNotFoundError: No module named 'manager_core'

Cannot import /data/config/comfy/custom_nodes/ComfyUI-Manager module for custom nodes: No module named 'manager_core'
```
Going to figure this out later

**Reply by gl-prout** [reply]
In `ComfyUI-Manager/glob/manage_server.py`

I changed the line
```python
import manager_core as core
```
into
```python
from . import manager_core as core
```

**Reply by raonsol** [reply]
Is this only workaround? I am facing the same issue, and it seems that it is related with PYTHONPATH.

**Reply by zzddwyff** [reply]
不错不错

**Reply by therealpygon** [reply]
This also fixed the issue for me when receiving the error "No module named 'manager_core'". 

Edit: That said, I have issue installing modules. I am beginning to wonder if the cause of the issue is not how ComfyUI is installed, but rather, that ComfyUI-Manager makes assumptions that it is only ever installed in comfy/custom_nodes and not in a custom_nodes folder set up by extra_model_paths, like when running inside a docker environment. Indeed, if I force the installation to be in the comfy custom_nodes directory, the manager works correctly.

**Reply by pixelpainter** [reply]
> In `ComfyUI-Manager/glob/manage_server.py`
> 
> I changed the line
> 
> ```python
> import manager_core as core
> ```
> 
> into
> 
> ```python
> from . import manager_core as core
> ```

Thank you! This fixed it for me on windows 11 with a NON standalone version

**Comment by nehemiascr** [comment]
I've just installed the manager in the comfy [Dockerfile](https://github.com/AbdBarho/stable-diffusion-webui-docker/blob/master/services/comfy/Dockerfile)

take a look at https://github.com/nehemiascr/stable-diffusion-webui-docker/commit/91f163bb879dfa23c9db3c70c72eb149c61f75de
```
RUN git clone https://github.com/ltdrdata/ComfyUI-Manager ${ROOT}/custom_nodes/ComfyUI-Manager && \
  pip install -r ${ROOT}/custom_nodes/ComfyUI-Manager/requirements.txt
```

**Reply by drakarah** [reply]
I tried this but the problem is that ComfyUI Manager won't see the other custom nodes installed the extra paths

**Comment by Meysmerized** [comment]
Thanks to everyone's input here
managed to make it work proper... or at least the red text claiming there are errors in the manager UI when I install new nodes don't change that it actually works.

Followed what gl-prout said

**1.** 
Created the startup.sh script in stable-diffusion-webui-docker/data/config/comfy 

```
#!/bin/bash

if [ -f "/data/config/comfy/custom_nodes/ComfyUI-Manager/requirements.txt" ]; then
  pip install -r /data/config/comfy/custom_nodes/ComfyUI-Manager/requirements.txt
fi
```


**2.** 
with the requirements.txt at the location it says, can't remember if it was already there or if I had to create it, but the key thing is from then on, whenever I get the ModuleNotFoundError: No module named 'numba' and the such, I just add numba (or the module name it says in between the '') close and restart. Haven't had problems since I started doing that.

I came to this from therealpygon talking about forcing installation

So my requirements.txt file looks like this now:

```
GitPython
PyGithub
matrix-client==0.4.0
transformers
huggingface-hub>0.20
typer
rich
typing-extensions
scikit-learn
numba
```

**3.**
also did the :

In ComfyUI-Manager/glob/manage_server.py

I changed the line
`import manager_core as core`
into
`from . import manager_core as core`


**note for people on the bottom of the learning curve (on the wrong side) like me:**
Terminal said: ModuleNotFoundError: No module named 'sklearn'
when I tried just copying sklearn into the requirement.txt list, the terminal told me The 'sklearn' PyPI package is deprecated, use 'scikit-learn' rather than 'sklearn' for pip commands.
I replaced sklearn with scikit-learn in the list, closed restarted and voila
So, it's worth checking the terminal and learning it's secrets when things don't work, til

_Thank you all for feeding discussions like this with lifesaving informations I usually only lurk for, truly grateful_

**Reply by thefoolofemmaus** [reply]
Hey! This method worked for me other than step 3. That resulted as an error, but Manager loaded as expected once I reverted that change out.
